### PR TITLE
cache-external-json-ajax-requests

### PR DESCRIPTION
**Pull request by hesedel**

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -149,7 +149,7 @@ const preloadResources = opt => {
     if (/^data:|blob:/i.test(responseUrl)) return;
     const ct = response.headers()["content-type"] || "";
     const route = responseUrl.replace(basePath, "");
-    if (/^http:\/\/localhost/i.test(responseUrl)) {
+    if (/^http:\/\/localhost/i.test(responseUrl) || ct.includes("json")) {
```
Comment (stereobooster) [inline index.js]: this would have other effects, except what you want. You will need to rewrite this branch of logic

Comment (hesedel) [reply index.js]: It is working like I want. All the logic under this block is just checking if the resource is an image/css/js/json file under localhost. So this one just matches with the json bit.

Comment (hesedel) [reply index.js]: unless we add another option to the config? cacheThirdPartyAjaxRequests?

Comment (hesedel) [reply index.js]: actually, I would assume at first tha thet `cacheAjaxRequests` option would already included 3rd-party requests (all ajax requests, not just local).

Comment (hesedel) [reply index.js]: I could also just move the json part outside of this `if localhost` block, into its own but just repeat `uniqueResources` stuff:

`
const responseUrl = response.url();
if (uniqueResources.has(responseUrl)) return;
...
uniqueResources.add(responseUrl);
`

Comment (stereobooster) [reply index.js]: Yes code will work right, I was confused a bit. The issue here is that code we lose clarity a bit.

before

```
if () {
// localhost
} else {
// 3rd-party
}
```

after

```
if () {
// localhost or 3rd-party
} else {
// 3rd-party
}
```

You see what I mean? I guess we can move out the condition about JSON

Comment (hesedel) [reply index.js]: Yes, you're right, I will make the change. 👌

File: index.js [hunk 2]
```diff
@@ -149,6 +149,12 @@ const preloadResources = opt => {
     if (/^data:|blob:/i.test(responseUrl)) return;
     const ct = response.headers()["content-type"] || "";
     const route = responseUrl.replace(basePath, "");
+    if (cacheAjaxRequests && ct.includes("json")) {
+      if (uniqueResources.has(responseUrl)) return;
+      const json = await response.json();
+      ajaxCache[route] = json;
+      uniqueResources.add(responseUrl);
```
Comment (stereobooster) [inline index.js]: Also we do not need to `preconnect` to third-party server if it is already cached.

Comment (stereobooster) [reply index.js]: The number of options leads to high number of possible cases, that is why I do not like to add options 😞. Zero-config is the best

**Comment by stereobooster** [comment]
Can you please, try to add test for it? Thanks

**Comment by hesedel** [comment]
I'm not sure how to add a test for this though. It's weird if I get a 3rd party json from some other place on the web.

**Comment by stereobooster** [comment]
You can try `http://127.0.0.1` as a 3-rd party website

**Comment by hesedel** [comment]
So I just tried `http://127.0.0.1` with a fixed port, it's not working, `http://localhost` works.

**Comment by stereobooster** [comment]
hm.. I will look into it later. The question is how server binded, but this should be possible

**Comment by hesedel** [comment]
I'm closing this – I just found about `snapSaveState` and I think that one suits better what I was trying to achieve – I want to save partial data from the ajax json response and not everything, in this case, translations, so as not to have a flash of untranslated text... I definitely don't want to save all the translations in every page.

Would still be curious though as to what IP address localhost resolves to.

**Pull request closed without merging**